pkg/net/net.go

```diff
@@ -74,6 +75,61 @@ func (o *Overlay) loop(ch <-chan EvtState, state event.Emitter) {
 			}
 		}
 	}()
+
+	for {
+		time.Sleep(500 * time.Millisecond) // TODO: decide optimal tick for updating neighbors
```

Comment:
There's a tradeoff between liveliness (how quickly a cluster reaches a steady-state, or repairs itself) and chattiness (how much bandwidth a cluster utilizes).  Ideally, a cluster that has converged on a steady-state will use bandwidth sparingly, and a cluster that has not yet converged will be more aggressive while using a bounded amount of bandwidth.

With this in mind, 500ms strikes me as a very chatty protocol.  The advantage is that we don't have to wait very long for the cluster to boot up, but I think we can finesse this a bit.  What do you think about the following approach?

A 500ms tick + capped exponential backoff when a node has fewer than `MaxSize()` neighbors, where the backoff resets each time the node finds a new peer.  The result is that a node with two modes of operation:  bootstrap and steady-state.  A node is in bootstrap mode when it has fewer than `MaxSize()` neighbors, else it is in steady-state mode.
- In **steady-state** mode, a node performs peer exchange at a fixed interval (say, every minute).
- In **bootstrap** mode, a peer begins by performing a peer exchange every 500ms.  Every time a peer exchange is successful (i.e. a new neighbor is discovered), the exponential backoff is reset (_i.e._ the next exchange will take place 500ms later).  Else, the neighbor exponentially backs off (up to the steady-state tick duration).

It may also be desirable to introduce some jitter to tick duration because it is often the case that many nodes are started simultaneously (e.g. via an Ansible script or something).  We want to spread push/pull events over time to avoid network storms.  [`jitterbug`](https://github.com/lthibault/jitterbug) can help with this, and incidentally, it can also help us implement the modes of operation described above.

What do you think of the following approach?  (Code not tested.)

```go
b := backoff{
    MinDuration:    time.Millisecond * 500,
    MaxDuration:   time.Minute,
    Factor: 2,         // double the duration with each backoff
}

ticker := jitterbug.New(
    time.Millisecond * 500,  // base tick interval
   &b,                                    // backoff would satisfy jitterbug.Jitter interface
)
defer ticker.Stop()

for range ticker.C {
    // ...

    err, n := o.setNeighborhod(o.ctx(), peers);
    if err != nil {
        if errors.Is(err, errNoNewPeers) {
            b.Reset()  // ignored if neighborhood is full
        }
    }

    // If true, b ignores calls to Reset and ensures that ticker.C fires every minute.
    b.SetSteadyState(n >= o.n.MaxSize())
}
```

Reply:
I find it a nice solution 👍 I will implement it then. I'm starting to notice you have written many unexpected libraries  🤔 you are a very prolific coder.

Reply:
👍 

> I'm starting to notice you have written many unexpected libraries 🤔 you are a very prolific coder.

Thanks!  I think I've built up a bit of an ecosystem of utilities over the years that I constantly reuse across projects.  Here are a few that we'll definitely be using (or are already using):

- https://github.com/lthibault/log
- https://github.com/lthibault/util
- https://github.com/lthibault/treap (one of my favorites; implemented specifically for Wetware)

Reply:
That's very useful 👍 I will look at them